roles/neutron_namespace_drain/tasks/drain-dhcp.yml

```
@@ -0,0 +1,93 @@
+  ansible.builtin.command: >
+    {{ neutron_namespace_drain_venv }}/bin/openstack
+    network agent list --host {{ neutron_namespace_drain_host }}
+    --agent-type dhcp -f value -c ID
+  register: src_dhcp_id
+  environment: "{{ openstack_auth_env }}"
+  changed_when: false
+  delegate_to: "{{ neutron_namespace_drain_host }}"
+  vars:
+    ansible_host: "{{ hostvars[neutron_namespace_drain_host].ansible_host }}"
+
+- name: Set fact containing SRC_DHCP_ID
+  ansible.builtin.set_fact:
+    dhcp_src_id: "{{ src_dhcp_id.stdout }}"
+
+- name: Get DHCP agent network IDs
+  ansible.builtin.command: >
+    {{ neutron_namespace_drain_venv }}/bin/openstack
+     network list --agent {{ dhcp_src_id }} -f value -c ID
+  register: dhcp_agent_ids
+  environment: "{{ openstack_auth_env }}"
+  changed_when: false
+  delegate_to: "{{ neutron_namespace_drain_host }}"
+  vars:
+    ansible_host: "{{ hostvars[neutron_namespace_drain_host].ansible_host }}"
+
+- name: Set DHCP agent network IDs
+  ansible.builtin.set_fact:
+    dhcp_agent_ids: "{{ dhcp_agent_ids.stdout_lines }}"
+
+- name: Remove DHCP agent
+  ansible.builtin.command: >
+    {{ neutron_namespace_drain_venv }}/bin/openstack
+    network agent remove network {{ dhcp_src_id }} {{ item }} --dhcp
+  environment: "{{ openstack_auth_env }}"
+  changed_when: false
+  delegate_to: "{{ neutron_namespace_drain_host }}"
+  vars:
+    ansible_host: "{{ hostvars[neutron_namespace_drain_host].ansible_host }}"
+  with_items:
+    - "{{ dhcp_agent_ids }}"
+
+- name: Wait for no more dhcp agents to be attached to the host
+  block:
+    - name: Retry count
+      ansible.builtin.set_fact:
+        retry_count: "{{ 0 if retry_count is undefined or retry_count == 'reset' else retry_count | int + 1 }}"
+        max_retries: 20
+
+    - name: Verify dhcp agents exist
+      ansible.builtin.command: >
+        {{ neutron_namespace_drain_venv }}/bin/openstack
+        network list --agent {{ dhcp_src_id }} -f value -c ID
+      environment: "{{ openstack_auth_env }}"
+      changed_when: false
+      delegate_to: "{{ neutron_namespace_drain_host }}"
+      vars:
+        ansible_host: "{{ hostvars[neutron_namespace_drain_host].ansible_host }}"
+      register: agent_status
+
+    - name: Fail if DHCP agent still attached
+      ansible.builtin.fail:
+        msg: a DHCP agent is still attached to "{{ l3_src_id }}"
+      when: agent_status.stdout | length > 0
+
+    - name: Reset retry count after success
+      ansible.builtin.set_fact:
+        retry_count: reset
+  rescue:
+    - name: Fail if maximum retries are exceeded
+      ansible.builtin.fail:
+        msg: |
+          Maximum retries waiting for DHCP agents to be detached reached
+      when: retry_count | int == max_retries
+
+    - name: Reset retry counter if max retries reached (exit loop)
+      ansible.builtin.set_fact:
+        retry_count: reset
+      failed_when: retry_count == 'reset'
+      when: retry_count | int >= max_retries | int
```

I'm not sure I really get what's going on here, but if I understand correctly, could we do this?

```suggestion
- name: Wait for no more dhcp agents to be attached to the host
  ansible.builtin.command: >
    {{ neutron_namespace_drain_venv }}/bin/openstack
    network list --agent {{ dhcp_src_id }} -f value -c ID
  environment: "{{ openstack_auth_env }}"
  changed_when: false
  delegate_to: "{{ neutron_namespace_drain_host }}"
  vars:
    ansible_host: "{{ hostvars[neutron_namespace_drain_host].ansible_host }}"
  register: agent_status
  failed_when: agent_status.stdout | length > 0
  retries: 20
  delay: 5
```

---

I think this should work given what we're trying to do, and looks a fair bit cleaner.